Upgrade and Checkout Page

garzeah opened the pull request:
### What this PR does (required):
- Created the front end for the Upgrade and Checkout page
- Fixed the log out bug

### Screenshots / Videos (required):
<img width="1792" alt="Screen Shot 2021-07-13 at 11 41 03 PM" src="https://user-images.githubusercontent.com/29414787/125575446-ab22a202-4470-4a1c-9523-0dc68bb7d09f.png">

![Kapture 2021-07-16 at 00 29 45](https://user-images.githubusercontent.com/29414787/125909722-58cec43e-7194-46ed-a79b-dec48e436b70.gif)


### Any information needed to test this feature (required):
- Log in, head to the dashboard and click the upgrade page

moffatethan commented on client/src/context/useAuthContext.tsx:
```diff
@@ -22,18 +22,9 @@ export const AuthProvider: FunctionComponent = ({ children }): JSX.Element => {
-    },
-    [history],
-  );
+  const updateLoginContext = useCallback((data: AuthApiDataSuccess) => {
```
Has there been a performance concern that prompts us to utilize `useCallback` here. By doing this there is more work being done behind the scenes.

garzeah replied:
Hey Ethan, this was a part of the starter code but I suspect it is because we're utilizing useEffect. If we were to remove it, I believe we would get infinite re-renders if we let useEffect handle it w/o a callback since state would be a dependency in useEffect.

moffatethan commented on client/src/pages/Checkout/Checkout.tsx:
```diff
@@ -0,0 +1,127 @@
+    cvc: '',
+  });
+
+  const months = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'];
+  const years = ['21', '22', '23', '24', '25', '26', '27', '28', '29', '30'];
```
While the months is okay, the years is not practical. I don't want to have to maintain this list of years for expiration date. I would suggest either refactoring and just letting the user input the format `MM/YY` or how years a bit more automated. 

Truthfully, you could just look into letting them input `MM/YY` and mask the input and then do further logic to make sure the card isn't already expired or let Stripe handle that logic.

garzeah replied:
Hey Ethan, I've decided to let the user input the MM/YY. Also, I only masked the CVC because the mock up doesn't mask the expiration date. However, if you want me to mask it, let me know! Lastly, I'm going to let Stripe handle that logic. Thanks for the feedback!

moffatethan commented on client/src/pages/Checkout/Checkout.tsx:
```diff
@@ -0,0 +1,127 @@
+    </MenuItem>
+  ));
+
+  return (
+    <Box className={classes.page}>
+      <Navigation />
+      <Box my={15} px={40} py={10} className={classes.checkoutBox}>
+        <Box className={classes.headerBox}>
+          <h1>Checkout</h1>
+        </Box>
+        <Box mb={2.5} className={classes.cardNumberBox}>
+          <h3>Card number</h3>
+          <FormControl>
+            <OutlinedInput
+              placeholder="7364 8257 2746 6153"
+              name="cardNumber"
+              value={cardInfo.cardNumber}
```
Should use Formik here, it's available in the project and can really make form work more maintainable and reliable.

moffatethan replied:
Also, Yup for our validation work

garzeah replied:
Hey Ethan, just converted the component to use formik and yup!

moffatethan commented on client/src/pages/Dashboard/EventModal/EventModal.tsx:
```diff
@@ -26,8 +26,6 @@ export default function EventModal({ open, setOpen }: Props): JSX.Element {
     setForm({ ...form, duration: e.target.value as string });
   };
 
-  console.log(form);
```
Thanks for removing that.

moffatethan reviewed (commented):
Left some comments for you to address, I think we should probably utilize Formik here as opposed to doing everything ourselves.

garzeah commented:
Hey Ethan, I've implemented your suggestions. Also, I ended up changing the structure and styling a bit to make the page more responsive. Lastly, I updated the gif!

moffatethan reviewed (approved):
Really good improvement Andy, good work! Thank you!
